File: go/yacht.go
package yacht

func Score(dice []int, category string) int {
	mp := map[int]int{}
	sum := 0
	for _, d := range dice {
		mp[d]++
		sum += d
	}

	switch category {
	case "yacht":
		for _, v := range mp {
			if v == 5 {
				return 50
			}
		}
		return 0
	case "ones":
		return mp[1]
	case "twos":
		return mp[2] * 2
	case "threes":
		return mp[3] * 3
	case "fours":
		return mp[4] * 4
	case "fives":
		return mp[5] * 5
	case "sixes":
		return mp[6] * 6
	case "full house":
		has3, has2 := false, false
		for _, v := range mp {
			if v == 3 {
				has3 = true
			} else if v == 2 {
				has2 = true
			}
		}
		if has3 && has2 {
			return sum
		}
		return 0
	case "four of a kind":
		for d, v := range mp {
			if v >= 4 {
				return d * 4
			}
		}
		return 0
	case "little straight":
		for _, v := range mp {
			if v != 1 {
				return 0
			}
		}
		if mp[1] == 1 && mp[6] == 0 {
			return 30
		}
		return 0
	case "big straight":
		for _, v := range mp {
			if v != 1 {
				return 0
			}
		}
		if mp[6] == 1 && mp[1] == 0 {
			return 30
		}
		return 0
	case "choice":
		return sum
	default:
		return 0
	}
}

File: lua/yacht.lua
local yacht = {}

function indexOf(t, val)
    for i, v in ipairs(t) do
        if v == val then
            return i
        end
    end
    return -1
end

function findCount(t, val)
    for _, v in pairs(t) do
        if v == val then
            return true
        end
    end
    return false
end

yacht.score = function(dice, category)
    table.sort(dice)

    local t = {}
    local sum = 0
    for _, v in ipairs(dice) do
        t[v] = (t[v] or 0) + 1
        sum = sum + v
    end

    local nums = { "ones", "twos", "threes", "fours", "fives", "sixes" }

    local index = indexOf(nums, category)
    if index ~= -1 then
        return (t[index] or 0) * index
    end

    if category == "yacht" then
        if findCount(t, 5) then
            return 50
        end
    end

    if category == "full house" then
        if findCount(t, 3) and findCount(t, 2) then
            return sum
        end
    end

    if category == "four of a kind" then
        for k, v in pairs(t) do
            if v >= 4 then
                return k * 4
            end
        end
    end

    if category == "little straight" then
        if table.concat(dice) == "12345" then
            return 30
        end
    end

    if category == "big straight" then
        if table.concat(dice) == "23456" then
            return 30
        end
    end

    if category == "choice" then
        return sum
    end

    return 0
end

return yacht
